### PR TITLE
Cholesky example

### PR DESCRIPTION
**Pull request by oxinabox**
@willtebbutt @duvenaud and I sat down and thought:
"You know what would be fun? To implement Cholesky in Dex.

This is as far as we got.

<s>Error seems to grow on the diagonal and we don't really know why</s>

### CODE REVIEW

File: examples/chol.dx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+reconst = for i. for j. psd.i.j - (mmp l (transpose l)).i.j
+
+:p reconst
+> [ [0.0, 0.0, 0.0, 0.0, -1.110223e-16, 1.110223e-16]
+> , [0.0, -8.155129e-3, 0.0, 0.0, 0.0, 0.0]
+> , [0.0, 0.0, -0.22620222, 0.0, 0.0, -1.110223e-16]
+> , [0.0, 0.0, 0.0, -0.847557, 0.0, -3.330669e-16]
+> , [-1.110223e-16, 0.0, 0.0, 0.0, -4.800562, 0.0]
+> , [1.110223e-16, 0.0, -1.110223e-16, -3.330669e-16, 0.0, -3.8064377] ]
```
Comment (oxinabox) [inline examples/chol.dx]: Just doesn't seem very accurate.
This whole matrix should be 0.0

File: examples/chol.dx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    innersum = sum for k. ..
+      select (inLower j k) (r.i.k * r.j.k) 0.0
+    (x.i.j - innersum) / r.j.j
+  ..
+  fs = ..
+  for outerrow::n. ..
+    lam r.
+    for i::n. ..
+      select (inUpper i outerrow) ..
+        (for j::n. r.i.j) .. -- copy
```
Comment (oxinabox) [inline examples/chol.dx]: If this doesn't optimize out, then it is an unfortunate increase in time complexity of the algorithm  over the mutating version.

File: examples/chol.dx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+
+
+' ## Cholesky Algorithm
+
+chol :: (n=>n=>Real) -> (n=>n=>Real)
+chol x = ..
+  diagOp r i = 
+    innersum = sum for k::n. ..
+      select (inLower k i) (sq r.i.k) 0.0
```
Comment (oxinabox) [inline examples/chol.dx]: This should be:
`select (inLower i k) (sq r.i.k) 0.0`

Comment (duvenaud) [reply examples/chol.dx]: Did that fix it?

Comment (oxinabox) [reply examples/chol.dx]: Not tested yet.

Comment (oxinabox) [reply examples/chol.dx]: that swap makes everything NaN, except first row.
But it is still correct by the formula AFAICT

File: examples/chol.dx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+chol :: (n=>n=>Real) -> (n=>n=>Real)
+chol x = ..
+  diagOp r j = 
+    innersum = sum for k::n. ..
+      select (inLower j k) (sq r.j.k) 0.0
+    sqrt (x.j.j - innersum)
+  lowerOp r i j = 
+    innersum = sum for k. ..
+      select (inLower j k) (r.i.k * r.j.k) 0.0
+    jdiag = diagOp r j -- can't look up r.j.j as it is not computed yet.
```
Comment (oxinabox) [inline examples/chol.dx]: this bit was the key thing we had wrong before.
We were reading the `r.j.j` before it is set to the correct value.
`r` is only done up to row `j-i` at this time.

So we need to recompute it here.
Hopefully dex can do Common Subexpression Elimnation and remove this.
But its still kinda ugly.

**Comment by oxinabox** [comment]
I feel like there is some kind of abstraction we need for structured arrays like this lowered triangular one.
And that it might be related to expressions about index types.
I think to do that one need to step away from thinking of arrays as collectioms of rows.
But if you do that you also have to step away from the idea that you can apply indexs one at a time to slice.

**Comment by dougalm** [comment]
Wow, great work!

> "You know what would be fun? To implement Cholesky in Dex.

Your idea of fun agrees with mine :)

> It would be nice to have either less than and greater than for indices, or ragged arrays.

Comparing indices with `<`/`==`/`>` should work already. I'll look into what's going wrong.

> Writing basic numerical algorithms efficiently without in-place mutation is going to be a challenge...

I agree and I think we should look into linear/uniqueness types so we can guarantee that updates will happen in-place. We already have one linear type system, why not a second? :). Short of that, I suspect we can get quite far with a best-effort buffer reuse optimization. That's all we have in JAX/XLA and it works well in practice. You can use `:flops` to see what the actual complexity is and check whether the optimization worked. I'm actually working on related optimizations for AD and I might have something working quite soon. I suggest we merge this PR for now, with a test that shows its O(N^4) behavior. Then we'll have the satisfaction of updating it to show O(N^3) once we add the optimization.

**Comment by oxinabox** [comment]
> Comparing indices with </==/> should work already. I'll look into what's going wrong.

Maybe it does and we were bracketing it wrong.
That happened a few times.

**Comment by oxinabox** [comment]
Will tryout `:flops` tomorrow and write that test.

**Comment by oxinabox** [comment]
Ok, this seems good to go to me.
Only thing is I can't check the flops are still correct as #57

**Comment by dougalm** [comment]
Right now this a good illustration of two of Dex's shortcomings.
  1. It doesn't work in-place, so the complexity is worse than O(N^3). This should now be fixable by writing it in terms of [`seq` ](https://github.com/google-research/dex-lang/commit/ea6557f57aa96a43d9f8192dbca69ff0dfdc4730) rather than `scan`. I'll take a crack at that and see how it looks.
  2. There's a lot of selecting against 0. We really just want to be iterating over a triangular domain, but Dex doesn't have a way to do that. This is something I'm actively working on. I think I have a solution, but I haven't tried implementing it yet.

Thanks for battling through these! I'll merge the current example and hopefully we'll come back and revise it soon.

**Pull request merged**